src/os/bluestore/BlueFS.cc

```diff
@@ -3038,54 +3041,38 @@ void BlueFS::_consume_dirty(uint64_t seq)
   }
 }
 
-// Extends log if its free space is smaller then bluefs_min_log_runway.
-// Returns space available *BEFORE* adding new space. Signed for additional <0 detection.
-int64_t BlueFS::_maybe_extend_log()
-{
+int64_t BlueFS::_extend_log(uint64_t amount, bool ignore_compaction) {
   ceph_assert(ceph_mutex_is_locked(log.lock));
-  // allocate some more space (before we run out)?
-  // BTW: this triggers `flush()` in the `page_aligned_appender` of `log.writer`.
-  int64_t runway = log.writer->file->fnode.get_allocated() -
-    log.writer->get_effective_write_pos();
-  if (runway < (int64_t)cct->_conf->bluefs_min_log_runway) {
-    dout(10) << __func__ << " allocating more log runway (0x"
-	     << std::hex << runway << std::dec  << " remaining)" << dendl;
-    /*
-     * Usually, when we are low on space in log, we just allocate new extent,
-     * put update op(log) to log and we are fine.
-     * Problem - it interferes with log compaction:
-     * New log produced in compaction will include - as last op - jump into some offset (anchor) of current log.
-     * It is assumed that log region (anchor - end) will contain all changes made by bluefs since
-     * full state capture into new log.
-     * Putting log update into (anchor - end) region is illegal, because any update there must be compatible with
-     * both logs, but old log is different then new log.
-     *
-     * Possible solutions:
-     * - stall extending log until we finish compacting and switch log (CURRENT)
-     * - re-run compaction with more runway for old log
-     * - add OP_FILE_ADDEXT that adds extent; will be compatible with both logs
-     */
-    if (log_forbidden_to_expand.load() == true) {
-      return -EWOULDBLOCK;
+  if (!ignore_compaction) {
+    std::unique_lock<ceph::mutex> ll(log.lock, std::adopt_lock);
+    while (log_forbidden_to_expand.load() == true) {
+      log_cond.wait(ll);
     }
-    vselector->sub_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
-    int r = _allocate(
+    ll.release();
+  }
+  vselector->sub_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
+  int r = _allocate(
       vselector->select_prefer_bdev(log.writer->file->vselector_hint),
-      cct->_conf->bluefs_max_log_runway,
+      amount, // FIXME: should this be split with max_run_way parts?
       0,
       &log.writer->file->fnode);
-    ceph_assert(r == 0);
-    vselector->add_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
-    log.t.op_file_update_inc(log.writer->file->fnode);
-  }
-  return runway;
+  ceph_assert(r == 0);
+  vselector->add_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
+  log.t.op_file_update_inc(log.writer->file->fnode);
```

Comment:
Here you add `UPDATE_INC(1,..)` on existing `log.t` transaction.
It will be showstopper for _replay(), since data will be located like this:
```
RUNWAYRUNWAYnewrunwaynewrunway
ORIGTRANSORIGTRANSupdateino1
```
Where update of bluefs log (ino.1) is on the part of disk that log was not aware that it is a part of log.
It must be something like this:
```
RUNWAYRUNWAYnewrunwaynewrunway
updateino1ORIGTRANSORIGTRANS
```

Reply:
done with last rebase